dot/state/grandpa.go

```diff
@@ -26,42 +37,477 @@ var (
+
+	for _, forced := range s.forcedChanges {
+		announcingHash := forced.announcingHeader.Hash()
+		effectiveNumber := forced.effectiveNumber()
+
+		if importedBlockHash.Equal(announcingHash) && effectiveNumber == importedBlockHeader.Number {
+			forcedChange = forced
+			break
+		}
+
+		isDescendant, err := s.blockState.IsDescendantOf(announcingHash, importedBlockHash)
+		if err != nil {
+			return fmt.Errorf("cannot check ancestry: %w", err)
+		}
+
+		if !isDescendant {
+			continue
+		}
+
+		if effectiveNumber == importedBlockHeader.Number {
+			forcedChange = forced
+			break
+		}
+	}
+
+	if forcedChange == nil {
+		return nil
+	}
+
+	forcedChangeHash := forcedChange.announcingHeader.Hash()
+	bestFinalizedNumber := forcedChange.bestFinalizedNumber
+
+	// checking for dependant pending scheduled changes
+	for _, scheduled := range s.scheduledChangeRoots {
+		if scheduled.change.effectiveNumber() > uint(bestFinalizedNumber) {
+			continue
+		}
+
+		scheduledBlockHash := scheduled.change.announcingHeader.Hash()
+		isDescendant, err := s.blockState.IsDescendantOf(scheduledBlockHash, forcedChangeHash)
+		if err != nil {
+			return fmt.Errorf("cannot check ancestry: %w", err)
+		}
+
+		if isDescendant {
+			return errPendingScheduledChanges
+		}
+	}
+
+	logger.Debugf("applying forced change: %s", forcedChange)
+
+	// send the telemetry s messages here
+	// afg.applying_forced_authority_set_change
+
+	currentSetID, err := s.GetCurrentSetID()
+	if err != nil {
+		return fmt.Errorf("cannot get current set id: %w", err)
+	}
+
+	err = s.setChangeSetIDAtBlock(currentSetID, uint(forcedChange.bestFinalizedNumber))
+	if err != nil {
+		return fmt.Errorf("cannot set change set id at block: %w", err)
+	}
+
+	newSetID, err := s.IncrementSetID()
+	if err != nil {
+		return fmt.Errorf("cannot increment set id: %w", err)
+	}
+
+	grandpaVotersAuthorities := types.NewGrandpaVotersFromAuthorities(forcedChange.nextAuthorities)
+	err = s.setAuthorities(newSetID, grandpaVotersAuthorities)
+	if err != nil {
+		return fmt.Errorf("cannot set authorities: %w", err)
+	}
+
+	err = s.setChangeSetIDAtBlock(newSetID, forcedChange.effectiveNumber())
+	if err != nil {
+		return fmt.Errorf("cannot set change set id at block")
+	}
+
+	logger.Debugf("Applying authority set forced change at block #%d",
+		forcedChange.announcingHeader.Number)
+
+	return nil
+}
+
+// forcedChangeOnChainOf walk through the forced change slice looking for
+// a forced change that belong to the same branch as blockHash parameter
+func (s *GrandpaState) forcedChangeOnChainOf(blockHash common.Hash) (forcedChange *pendingChange, err error) {
+	for _, change := range s.forcedChanges {
+		changeHeader := change.announcingHeader
+
+		var isDescendant bool
+		isDescendant, err = s.blockState.IsDescendantOf(
+			changeHeader.Hash(), blockHash)
+
+		if err != nil {
+			return nil, fmt.Errorf("cannot verify ancestry: %w", err)
+		}
+
+		if !isDescendant {
+			continue
+		}
+
+		return change, nil
```

Comment:
nit flip logic around in the for loop

```suggestion
		if isDescendant {
			return change, nil
		}		
```